ibis/expr/rewrites.py

```diff
@@ -82,6 +82,26 @@ def rewrite_dropna(_):
     return ops.Filter(_.parent, tuple(preds))
 
 
+@replace(p.StringSlice)
+def rewrite_stringslice(_, **kwargs):
+    """Rewrite StringSlice in terms of Substring."""
+    if _.end is None:
+        return ops.Substring(_.arg, start=_.start)
+    if _.start is None:
+        return ops.Substring(_.arg, start=0, length=_.end)
+    if (
+        isinstance(_.start, ops.Literal)
+        and isinstance(_.start.value, int)
```

Comment:
The rewrite rule is good as is, just noting that this could be spelled as a pattern as well: `p.StringSlice(p.Literal(int), p.Literal(int)`.

Reply:
As in there would be a different rule for each of the patterns? (Literal, None), (None, literal), (literal, literal)? Maybe a full example of how this could be spelled? If there are different rules then would I need to import three different rewrites in all the compilers?

I'd love to learn for next time.

Reply:
I think this is ideal in this current scenario, just wanted to highlight you that nested patterns are also supported. 

There are various ways to spell this, here is a more complicated one showing what can be done with the rewrite system:

```py
from ibis.expr.rewrites import p, d

p.StringSlice(end=None) >> d.Substring(_.arg, start=_.start) |
p.StringSlice(start=None) >> d.Substring(_.arg, start=0, length=_.end) |
p.StringSlice(p.Literal(int), p.Literal(int)) >> d.Substring(_.arg, start=_.start, length=_.end.value - _.start.value) |
p.StringSlice >> d.Substring(_.arg, start=_.start, length=d.Subtract(_.end, _.start)
```